### PR TITLE
virtcontainers: Extract shim related operations to a separate package

### PR DESCRIPTION
**Pull request by zsc-hw**
With the increase of code in the "virtcontainers" package,
 code coupling  is also increasing. Therefore, I want to
 recfactor the code of shim, agent, proxy, network
 and other related modules, decouple them and put
 them into a separate package.
    Now I put the relevant code of "shim" into the independent
 "shim" package. I hope it can be perfected through
 your suggestions. If possible, I will continue to reconstruct
 agent, proxy, network and so on.

Fixes: #890

Signed-off-by: Shucheng Zhang  <zhangshucheng@huawei.com>

### CODE REVIEW

File: virtcontainers/pkg/oci/utils.go [hunk 1]
```diff
@@ -22,6 +22,8 @@ import (
 	"github.com/sirupsen/logrus"
 
 	vc "github.com/kata-containers/runtime/virtcontainers"
+	shim "github.com/kata-containers/runtime/virtcontainers/shim"
+	vtypes "github.com/kata-containers/runtime/virtcontainers/types"
```
Comment (jshachm) [inline virtcontainers/pkg/oci/utils.go]: why jump around `shim` ,`vshim`, `types`,`vtypes` ？

Comment (zsc-hw) [reply virtcontainers/pkg/oci/utils.go]: Thank you for your question.

Now it has been unified into vtypes and vshim.

vtypes "github.com/kata-containers/runtime/virtcontainers/types"

vshim "github.com/kata-containers/runtime/virtcontainers/shim"

**Comment by sboeuf** [comment]
@zsc-hw 
Hi, I just got a look at your PR and I have several comments:
- What's the benefit of moving the shim interface to its own package? I feel that decoupling too much will lead to some complexity that's not really worth it.
- Moving the types to a separate and global package might be a good idea, but if we do so, please move those to `virtcontainers/pkg/types`.
- You didn't modify the unit tests yet, but that's where you might have some trouble with all those new packages.
- Please make sure your commit message complies with Kata Containers guidelines.

/cc @bergwolf @WeiZhang555 @egernst @amshinde

**Comment by WeiZhang555** [comment]
@sboeuf 

> What's the benefit of moving the shim interface to its own package? I feel that decoupling too much will lead to some complexity that's not really worth it.

It will lead to some complexity, but I think that's worth it. You know, I mentioned a lot that I hope to refactor the virtcontainers and decouple each components to seperate dirs, this requires lots of work, but I'm so lazy so this has been postponed for long time. I remember that @bergwolf mentioned this too.

Two benefits for seperating these dirs:

1. More friendly for scalability. Shim may not be a good example, take Hypervisor as an example, currently we only support qemu, we may want to support more hypervisors such as xen/libvirt/hyperV, putting them in seperate dir can make programmers easier to understand how to extend this.

2. Better readability. hierarchical codes are more friendly to code readers.

**Comment by bergwolf** [comment]
@sboeuf I tend to agree with @WeiZhang555. Modules with clear interfaces can be put in separate packages and it makes code maintenance easier IMHO.

**Comment by sboeuf** [comment]
@bergwolf @WeiZhang555 
If you really think this would be easier to understand for other developers, then I'm fine with that :)

**Comment by bergwolf** [comment]
Thanks @sboeuf ! The refactoring process might be slow but I think it is the right direction.

**Comment by egernst** [comment]
@sboeuf can you take another look at this PR?

**Comment by sboeuf** [comment]
@egernst yes I will!

**Comment by jodh-intel** [comment]
Ping @zsc-hw - do you think you'll be able to work on reformatting the commit message?

Also, it feels a bit odd having the `*_shim_test.go` files under `virtcontainers/` but the implementations in `virtcontainers/shim/*.go` - shouldn't the tests be moved too?

**Comment by egernst** [comment]
@WeiZhang555 @zsc-hw - I'd love to see this PR move forward -- any chance for updates here?

**Comment by zsc-hw** [comment]
@jodh-intel @egernst 
I'll update it these days.

**Comment by caoruidong** [comment]
penguins

**Comment by jodh-intel** [comment]
@zsc-hw - thanks for explaining. Please could you split this into 2 commits as requested by @sboeuf?

**Comment by raravena80** [comment]
@zsc-hw ping any updates? Thx!

**Comment by shuchengzg** [comment]
@raravena80
It will be updated in the next few days

**Comment by egernst** [comment]
friendly ping :)

**Comment by jodh-intel** [comment]
Branch needs updating due to conflicts.

**Comment by grahamwhaley** [comment]
@sameo - ptal, as afaik you are also looking at and undertaking much vc refactor and breakout.
@zsc-hw - looks like a major rebase is needed at least.

**Comment by jodh-intel** [comment]
Hi @zsc-hw - please could you rebase to resolve the conflicts on this PR?

**Pull request closed without merging**